BugFix: attempt to solve copy selection problem

SlySven opened the pull request:
We have established that the Copy as HTML and Copy as Image modes do not seem to have the problem that the plain Copy (as Text) one has. It is noticeable that the first two have the same process to select the portion of the `TBuffer` that is to be copied and that seems to be different to that of the third. So this changes the latter to use the same system that has been refactored out of the other two.

Also, since all the text we need is already in an arrangement that we can actually copy quite fast (a `QStringList`) there is no need to do a slow `QChar` by `QChar` copy of every `QChar` in the whole selection - instead we can take a copy of that section of the `TBuffer::lineBuffer` that contains the wanted lines - and then just trim off the leading `QChar`s on the first line that precede the selection stop (and replace them with spaces if it is a multi-line selection) and chop off the unwanted trailing `QChar`s on the last line... This way of doing things also avoids the complication that empty intermediate lines might not have ANY characters in them!

Also - I noted that the code in the Copy as HTML and Copy as Image that was similar to the code in the existing `normalisedSelection()` was actually faulty as when the start and end points were on the same line
(i.e. the Y values were the same), the test to see whether the points had to be swapped had an incorrect test for the X values also being the same, whereas they should have been the end point X value being less than the starting X value - as they are in the normalising method!

This should close #3285 (and possibly some other issues).

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: attempt to solve copy selection problem

We have established that the Copy as HTML and Copy as Image modes do not
seem to have the problem that the plain Copy (as Text) one has. It is
noticeable that the first two have the same process to select the portion
of the `TBuffer` that is to be copied and that seems to be different to
that of the third. So this changes the latter to use the same system that
has been refactored out of the other two.

Also, since all the text we need is already in an arrangement that we can
actually copy quite fast (a `QStringList`) there is no need to do a slow
`QChar` by `QChar` copy of every `QChar` in the whole selection - instead
we can take a copy of that section of the `TBuffer::lineBuffer` that
contains the wanted lines - and then just trim off the leading `QChar`s
on the first line that precede the selection stop (and replace them with
spaces if it is a multi-line selection) and chop off the unwanted trailing
`QChar`s on the last line... This way of doing things also avoids the
complication that empty intermediate lines might not have ANY characters in
them!

Also - I noted that the code in the Copy as HTML and Copy as Image that
was similar to the code in the existing `normalisedSelection()` was
actually faulty as when the start and end points were on the same line
(i.e. the Y values were the same), the test to see whether the points had
to be swapped had an incorrect test for the X values also being the same,
whereas they should have been the end point X value being less than the
starting X value - as they are in the normalising method!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Is `Select All` working all reliably for you?

Eraene commented:
Select all seems to be unreliable for some reason, but I'm not sure why. Sometimes it takes a few tries to work. It seems to lose reliability after selecting other text normally.

Kebap commented:
I never use select all, it seems quite pointless really, and now I look at it, it hardly ever works correctly. I tested about 20 times and it worked alright in 2-3 tries..!
edit: comment was not for this PR, but for PTB-05-29

SlySven commented:
IIRC Select all is missing something that updates the screen (to show everything highlighted) so that doesn't tend to happen until a scroll or something else causes a refresh...

SlySven commented:
Ah, the `TextEdit::slot_selectAll()` is missing a `normaliseSelection()` which will transfer the manually set `mDragStart` and `mDragSelectionEnd` `QPoint` values to `mPA` and `mPB` - although *then* the highlighting that is applied end-to-end does not get cleared when you click on somewhere to change/start a new selection... :hammer_and_wrench:

SlySven committed:
BugFix: get select all working in TConsole/TTextEdit

It needed a `normalisedSelection()` call to transfer the values manually
assigned to `(QPoint) mDragStart` and `mDragSelectionEnd` to the `mPA` and
`mPB` ones that are used when iterating through the selection.

This then showed that it was necessary to force an update of BOTH
`TTextEdit` instances for a `TConsole` so that a selection made in one
was shown in both if they both contain selected text - which is bound to
happen if ALL the text is selected.

A further click should *clear* the selection of ALL the text and start
a new drag operation.  This is now more likely to happen, but it does not
seem totally totally correct to me yet.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (approved):
Looking solid :+1:

njs50 commented:
@vadi2 should i be concerned about this seg fault? looks like it may be related?
[Mudlet_2020-06-01-125847_Nicks-MacBook-Pro.txt](https://github.com/Mudlet/Mudlet/files/4712770/Mudlet_2020-06-01-125847_Nicks-MacBook-Pro.txt)

vadi2 commented:
Any segfault is concerning, they should never happen. This one does seem related, @SlySven could you have a look?

vadi2 commented:
The seg fault I imagine happens under specific conditions and not just always whenever you copy right?

njs50 commented:
yeah. i haven't seen it again. not sure what exactly happened, i was probably mashing keys randomly. maybe hit ctrl-c after the selection had been purged from the scrollback or something silly like that. fwiw, that was a fresh build of develop as at a couple of hours ago. i'll keep an eye out for more.

SlySven commented:
:thinking: The bleeding edge of the backtrace is:
```
0   org.qt-project.QtCore         	0x000000010d5e902e QString::mid(int, int) const + 30
1   com.yourcompany.Mudlet        	0x000000010b549aa7 TTextEdit::getSelectedText(QChar) + 359
2   com.yourcompany.Mudlet        	0x000000010b5409e3 TTextEdit::slot_copySelectionToClipboard() + 35
3   com.yourcompany.Mudlet        	0x000000010b415366 TCommandLine::event(QEvent*) + 2182
...
```
That is looking like a `QString::mid(...)` was called on something that it shouldn't have - and for it to seg. fault the only thing I can thing is that it is specifically the FIRST line of a selection that had no characters in it - and looking at the mechanics I think that would have to be the multi-line selection case...

Humm, even when experimenting with a block of text with empty lines at the beginning and end I cannot reproduce this...

... so if it is happening I can't see how - (yet)!

vadi2 commented:
> fault the only thing I can thing is that it is specifically the FIRST line of a selection that had no characters in it 

This is very possible and this was the cause of the selection 'breaking' before, because of this check that seems to have been removed:

```cpp
    if(!mpBuffer->buffer[0].size()) {
        return text;
```

Without this check Mudlet crashed, with this check, selection just didn't work.

So I suspect the issue is still there, we just undid the check so now it'll crash instead, and this must be fixed asap.
